### PR TITLE
Get extra information about a route for a documentation 

### PR DESCRIPTION
**Pull request by bgaifullin**
Add method `GetPathRegexp`  which returns original regex for path.
Add method `GetMethods` which returns all available methods for path.

### CODE REVIEW

**Comment by bgaifullin** [comment]
Folks, please review.

**Comment by kisielk** [comment]
Can you give an example of where this will be used?

**Comment by bgaifullin** [comment]
We have service, where each deamon should register its API.
I automated this procedure by using router.Walk, but I do not have enough information about route.
currently I use custom function pathTemplate2Regexp, which converts the path template back to regex, and I use fact, that my service has only "GET" methods. It is temporary solution and I want this PR to make normal solution.

```go
// getResources returns the list of resources in format {"name", "action", "method", "pattern"}
func getResources(router *mux.Router) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		resources := []map[string]string{}

		router.Walk(func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
			path, err := route.GetPathTemplate()
			if err != nil {
				return err
			}
			name := route.GetName()
			// private urls starts with '_'
			if name != "" && name[0] != '_' {
				// TODO need interface to return methods from route.
				resource := map[string]string{
					"resource": name,
					"action":   "READ",
					"method":   "GET",
					"pattern":  pathTemplate2Regexp(path),
				}
				resources = append(resources, resource)
			}
			return nil
		})
		formatters["json"].Write(w, 200, resources, true)
	}
}
```

**Comment by kisielk** [comment]
Ok. Seems reasonable to me. @elithrar what do you think?

File: README.md [hunk 1]
```diff
@@ -189,16 +190,18 @@ func handler(w http.ResponseWriter, r *http.Request) {
 
 func main() {
     r := mux.NewRouter()
-    r.HandleFunc("/", handler)
-    r.HandleFunc("/products", handler)
+    r.Methods("GET").HandleFunc("/", handler)
```
Comment (elithrar) [inline README.md]: Don't modify these - should show both the simple (non-method) and precise (provide a method) use-cases.

File: README.md [hunk 2]
```diff
@@ -189,16 +190,18 @@ func handler(w http.ResponseWriter, r *http.Request) {
...
     r.HandleFunc("/articles", handler)
-    r.HandleFunc("/articles/{id}", handler)
+    r.Methods("GET", "PUT").HandleFunc("/articles/{id}", handler)
     r.Walk(func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
         t, err := route.GetPathTemplate()
         if err != nil {
             return err
         }
-        fmt.Println(t)
+        p, _ := route.GetPathRegexp()
```
Comment (elithrar) [inline README.md]: Don't ignore errors in READMEs; newbies learn bad practices ;)

File: route.go [hunk 3]
```diff
@@ -558,6 +558,36 @@ func (r *Route) GetPathTemplate() (string, error) {
...
+// GetPathRegexp returns the expanded regexp used to match route path.
+// This is useful for building simple REST API documentation and for instrumentation
+// against third-party services.
+// An error will be returned if the route does not define a path.
+func (r *Route) GetPathRegexp() (string, error) {
+	if r.err != nil {
+		return "", r.err
+	}
+	if r.regexp == nil || r.regexp.path == nil {
+		return "", errors.New("mux: route doesn't have a path")
```
Comment (elithrar) [inline route.go]: nitpick: s/doesn't/does not/

File: route.go [hunk 4]
```diff
@@ -558,6 +558,36 @@ func (r *Route) GetPathTemplate() (string, error) {
...
+func (r *Route) GetMethods() ([]string, error) {
+	if r.err != nil {
+		return nil, r.err
+	}
+	for _, m := range r.matchers {
+		if methods, ok := m.(methodMatcher); ok {
+			return []string(methods), nil
+		}
+	}
+	return nil, errors.New("mux: route doesn't have methods")
```
Comment (elithrar) [inline route.go]: nitpick: s/doesn't/does not/

**Comment by bgaifullin** [comment]
@elithrar Thank you for review. I have addressed your comments. Could you please re-review.

**Comment by bgaifullin** [comment]
I have just updated commits by using flag 'force', to avoid extra commits in log, which is not related to main purpose of this PR.
Is it OK, that GitHub still shows "changes requested" ?

**Comment by bgaifullin** [comment]
@elithrar, could you please look this patch again,  I really need this PR.

**Comment by adelowo** [comment]
I really need the `GetMethods`

File: README.md (hunk 1 above)
Comment (bgaifullin) [inline README.md]: done

File: README.md (hunk 2 above)
Comment (bgaifullin) [inline README.md]: done

File: route.go (hunk 3 above)
Comment (bgaifullin) [inline route.go]: done

File: route.go (hunk 4 above)
Comment (bgaifullin) [inline route.go]: done

**Comment by bgaifullin** [comment]
ping

**Comment by bgaifullin** [comment]
Is there any chance that this PR will be merged?

**Comment by elithrar** [comment]
I will review it today.

File: README.md [hunk 5]
```diff
@@ -190,15 +191,23 @@ func handler(w http.ResponseWriter, r *http.Request) {
...
+    r.Methods("POST").HandleFunc("/products", handler)
+    r.Methods("GET").HandleFunc("/articles", handler)
+    r.Methods("GET", "PUT").HandleFunc("/articles/{id}", handler)
     r.Walk(func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
         t, err := route.GetPathTemplate()
         if err != nil {
             return err
         }
-        fmt.Println(t)
+        p, err := route.GetPathRegexp()
```
Comment (elithrar) [inline README.md]: Add comments to describe *why* these are useful in the example - e.g. example output. The paths/regexp here is simple.

File: route.go [hunk 6]
```diff
@@ -558,6 +558,36 @@ func (r *Route) GetPathTemplate() (string, error) {
 	return r.regexp.path.template, nil
 }
 
+// GetPathRegexp returns the expanded regexp used to match route path.
```
Comment (elithrar) [inline route.go]: "expanded regular expression" (less abbreviations)

File: route.go [hunk 7]
```diff
@@ -558,6 +558,36 @@ func (r *Route) GetPathTemplate() (string, error) {
...
+	if r.err != nil {
+		return "", r.err
+	}
+	if r.regexp == nil || r.regexp.path == nil {
+		return "", errors.New("mux: route does not have a path")
+	}
+	return r.regexp.path.regexp.String(), nil
+}
+
+// GetMethods returns the methods used to build the methods match.
```
Comment (elithrar) [inline route.go]: "GetMethods returns the methods the route matches against."

**Comment by bgaifullin** [comment]
@elithrar thank you for review. I have addressed your comments.

**Comment by bgaifullin** [comment]
@elithrar, could you please review this PR again?

**Comment by elithrar** [comment]
LGTM. @kisielk?

**Comment by adelowo** [comment]
Ha, thank you..

**Comment by bgaifullin** [comment]
Thank you.

**Pull request merged**